File: 백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		
		for (int i = 0; i < n; i++) {
			String[] arr = new String[n];
			int size = 0;
			String s = sc.next();
			String[] temp = s.split("");
			for (String a : temp) {
				if (a.equals("(")) {
					arr[size++] = a;
				} else if (a.equals(")") && size!=0 && arr[size-1].equals("(")) {
						size--;
				} else {
					arr[size++] = a;
				}
			}
			if (size > 0) {
				System.out.println("NO");
			} else {
				System.out.println("YES");
			}
			
		}
	}
}

File: 백준/Silver/9012. 괄호/괄호.py
n = int(input())

for _ in range(n):
    arr = [0]
    a = input()
    for i in a:
        if arr[-1] != 0 and arr[-1] == '(' and i == ')':
            arr.pop(-1)
        else:
            arr.append(i)
    
    if arr[-1] != 0:
        print("NO")
    else:
        print("YES")